### PR TITLE
feat: add compress package & the ability to compress steps in a resolution

### PR DESCRIPTION
**Pull request by orandin**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

Feature

* **What is the current behavior?** (You can also link to an open issue here)

The steps aren't compressed.

* **What is the new behavior (if this is a feature change)?**

Now, we can compress the steps in a resolution to reduce the row size in database.
The compression algorithm used is saved in database to ensure backward compatibility in the case where the compression algorithm is changed. 

By default, no compression algorithm is used.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

No

* **Other information**:

To compress in `gzip`, update `utask-cfg` to add: 

```json
{
    "steps_compression": "gzip"
}
```

### CODE REVIEW

**Review by rbeuque74 — APPROVED** [review]
Thanks, that a nice proposal.
Will run the CI tests on Monday

**Comment by ovh-cds** [comment]
CDS Report test#1448.0 ✘  
* tests and lint
  * golangci-lint ✔ 
  * unit-test and integration-tests ✘

File: pkg/compress/compress.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+// RegisterCompression registers a custom compression algorithm.
+func RegisterCompression(name string, c Compression) error {
+	if c == nil {
+		return nil
+	}
+	compressionsMut.Lock()
+	defer compressionsMut.Unlock()
+	_, ok := compressions[name]
+	if ok {
```
Comment (rclsilver) [inline pkg/compress/compress.go:26]: Returning an error when `ok` is `true` is disturbing me. Maybe you could rename to `found` or `exists`?

File: pkg/compress/compress.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	compressionsMut sync.Mutex
+)
+
+type Compression interface {
+	Compress([]byte) ([]byte, error)
+	Decompress([]byte) ([]byte, error)
+}
+
+// RegisterCompression registers a custom compression algorithm.
+func RegisterCompression(name string, c Compression) error {
```
Comment (rclsilver) [inline pkg/compress/compress.go:19]: Maybe `RegisterAlgorithm`?

File: pkg/compress/compress.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+	return nil
+}
+
+func Get(name string) (Compression, error) {
+	compressionsMut.Lock()
+	defer compressionsMut.Unlock()
+
+	c, ok := compressions[name]
+	if !ok {
+		return nil, fmt.Errorf("no compression for type '%s'", name)
```
Comment (rclsilver) [inline pkg/compress/compress.go:39]: %s compression algorithm not found

File: pkg/compress/gzip/gzip.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+import (
+	"bytes"
+	"compress/gzip"
+	"io"
+
+	"github.com/ovh/utask/pkg/compress"
+)
+
+const CompressionName = "gzip"
```
Comment (rclsilver) [inline pkg/compress/gzip/gzip.go:11]: AlgorithmName or just Name ?

File: sql/migrations/009_resolution_compression_step.sql [hunk 5]
```diff
@@ -0,0 +1,11 @@
+-- +migrate Up
+
+ALTER TABLE "resolution" ADD COLUMN "compression_steps" TEXT;
```
Comment (rclsilver) [inline sql/migrations/009_resolution_compression_step.sql:3]: steps_compression_alg

File: utask.go [hunk 6]
```diff
@@ -123,6 +130,7 @@ type Cfg struct {
 	DashboardAPIPathPrefix                     string                   `json:"dashboard_api_path_prefix"`
 	DashboardSentryDSN                         string                   `json:"dashboard_sentry_dsn"`
 	EditorPathPrefix                           string                   `json:"editor_path_prefix"`
+	StepsCompression                           string                   `json:"steps_compression"`
```
Comment (rclsilver) [inline utask.go:133]: StepsCompressionAlg(orithm)

**Review by rclsilver — CHANGES_REQUESTED** [review]

**Review by rclsilver — CHANGES_REQUESTED** [review]
Could you fix the version in the schema.sql please?

I made some tests  and I guess that encrypting a gzip version of the json is more efficient:
```
root@939d37decdb7:/go/src/github.com/ovh/utask/cmd/test-compress# go run test-compress.go 
INFO[0000] file: json-1.json: 9581 bytes - gzip(encrypt(json)): 9637 bytes - encrypt(gzip(json)): 2626 bytes 
root@939d37decdb7:/go/src/github.com/ovh/utask/cmd/test-compress# 
```

**Comment by rbeuque74** [comment]
Last comment of @rclsilver seems relevant, we will gain more space doing it the other way :+1:

**Review by rclsilver — APPROVED** [review]

**Comment by ovh-cds** [comment]
CDS Report test#1463.0 ✘  
* tests and lint
  * golangci-lint ✘  
  * unit-test and integration-tests ✔

**Review by rclsilver — DISMISSED** [review]

**Review by rclsilver — APPROVED** [review]

**Review by rbeuque74 — APPROVED** [review]

**Pull request merged**